전남대 BE_조현서 6주차 과제 (1단계)

hxunsc opened the pull request:
- [팀 API 명세서](https://alive-tail-1fa.notion.site/API-778e7fee2b6c45f4bf19c06ac1e15461)
- 팀 API 명세서를 결정하면서 에러 코드를 지정하여 반환하도록 결정하였고 이로 인해 기존에 처리되지 않았던 부분까지 포함하여 파일 변경이 상당히 많습니다..
- 모든 테스트 코드가 통과되도록 수정을 완료하였습니다.
- 주문 기능과 관련하여 변경사항이 있습니다.
    - 저번주에는 카카오 토큰을 이용하여 '나에게 보내기' 기능을 구현하였습니다. 그러나 프론트엔드 팀과 논의 결과, 현재 소셜 로그인 기능이 없다는 점을 확인하였고 팀원 모두 주문하기 시 기존 토큰을 사용하여 구현한다고 하여, 주문 시 사용되는 토큰을 기존 로그인에서 발급되는 토큰으로 변경하였습니다.
- 해당 변경으로 인해 주문하기 기능 내의 '나에게 메시지 보내기' 기능에서 오류가 발생하여 현재는 이 부분을 임시로 주석 처리해 두었습니다.
- 이 문제에 대해 카카오 토큰과 기존 토큰을 모두 헤더에서 받는 방법을 고려해보았으나, 팀 간 합의된 명세서에 인증 토큰은 하나만 받기로 되어 있어 이 방법은 적절하지 않다고 판단하였고, 현재로서는 다른 해결 방안을 찾지 못하고 있는 상황입니다. 이 문제에 대한 조언 부탁드립니다..
- 응답 형식에 따라서 DTO가 많이 늘었는데, 너무 많아진 것 같기도 해서.. 이 부분은 상관이 없을까요?

hxunsc committed:
feat: Add CustomException

hxunsc committed:
refactor: API 명세서에 따른 카테고리 수정

hxunsc committed:
refactor: 에러 코드 반환 순서 지정

hxunsc committed:
refactor: API 명세서와 관련하여 멤버 코드 수정

hxunsc committed:
refactor: API 명세서에 따른 상품 생성 수정

hxunsc committed:
refactor: API 명세서에 따른 상품조회,수정,삭 제수정

hxunsc committed:
refactor: quantity -> StockQuantity 로 변수명 수정

hxunsc committed:
refactor: API 명세서에 따른 위시 수정

hxunsc committed:
refactor: API 명세서에 따른 상품 옵션 수정

hxunsc committed:
refactor: 토큰 관련 예외처리 추가

hxunsc committed:
refactor: API 명세서에 따른 주문 관련 수정

hxunsc committed:
refactor: 예외 처리 수정

hxunsc committed:
refactor: "Bearer " 추가

hxunsc committed:
refactor: 예외처리 수정

hxunsc committed:
refactor: 회원 응답 형식 수정

hxunsc committed:
refactor: WebProductController 수정

hxunsc committed:
feat: 옵션 수량 차감 기능 복구

hxunsc committed:
refactor: 팀 API 명세서에 따른 카테고리 수정

hxunsc committed:
refactor: 팀 API 명세서에 따른 상품 수정

hxunsc committed:
refactor: 팀 API 명세서에 따른 위시리스트 수정

hxunsc committed:
refactor: 팀 API 명세서에 따른 상품 옵션 수정

hxunsc committed:
refactor: 팀 API 명세서에 따른 주문 수정

hxunsc committed:
chore: Update Step1 Readme.md

hxunsc committed:
test: 테스트 코드 수정

repository와 service 수정으로 인한 테스트 코드 수정

hxunsc committed:
style: 코드 스타일 정렬

hxunsc committed:
style: 사용하지 않는 import문 삭제

seungy0 commented:
> 주문 기능과 관련하여 변경사항이 있습니다.
> 저번주에는 카카오 토큰을 이용하여 '나에게 보내기' 기능을 구현하였습니다. 그러나 프론트엔드 팀과 논의 결과, 현재 소셜 로그인 기능이 없다는 점을 확인하였고 팀원 모두 주문하기 시 기존 토큰을 사용하여 구현한다고 하여, 주문 시 사용되는 토큰을 기존 로그인에서 발급되는 토큰으로 변경하였습니다.

여기서 기존 토큰이 어떤걸 의미하는 걸까요? jwt를 말씀하시는 것으로 추측됩니다.

우선 인증과 '나에게 보내기'는 분리된 기능입니다.
일반적으론 OAuth access token을 서버에 저장할 필요가 없습니다.
하지만 '나에게 보내기'를 위해선 필요할 수 있겠네요.

jwt를 사용해 인증을 완료하면 서버에서 해당 유저에 대한 정보를 알 수 있을 것 입니다.
이 때 유저의 정보를 토대로 '나에게 보내기'를 하면 되는데, 이를 위해 access token이 필요하게 됩니다.
저였다면 refresh_token을 유저 테이블에 저장하는 식으로 구현했을 것 같습니다. 지금 상황에서는 가장 간편한 방법일 것 같네요.

seungy0 commented on src/main/java/gift/common/exception/GlobalExceptionHandler.java:
👍

hxunsc committed:
refactor: 주문하기 중 나에게 메세지 보내기 수정

멤버에 refresh_token을 추가하였고, 카카오api에서 이메일을 받아올 수
없어서 임시 이메일을 만들어서 jwt 추출하고 해당 jwt로 주문하기 시
메세지와 주문 둘 다 되도록 수정하였습니다.

hxunsc commented:
조언해주신 내용 기반으로 코드 수정했습니다!

seungy0 commented on src/main/java/gift/controller/AdminCategoryController.java:
```diff
@@ -37,27 +32,29 @@ public AdminCategoryController(CategoryService categoryService) {
-        @ApiResponse(responseCode = "400", description = "잘못된 요청"),
-        @ApiResponse(responseCode = "404", description = "카테고리를 찾을 수 없음")
+        @ApiResponse(responseCode = "204", description = "카테고리 수정 성공"),
+        @ApiResponse(responseCode = "-40402", description = "해당 카테고리를 찾을 수 없음"),
+        @ApiResponse(responseCode = "-40902", description = "카테고리 추가/수정 시 이름이 기존 카테고리와 중복되는 경우")
```
의도하신 바는 알겠으나, 실제론 코드가 500이라서, API 명세서 수정이 필요합니다.
<img width="1423" alt="스크린샷 2024-08-03 오후 3 26 48" src="https://github.com/user-attachments/assets/e7f95f6e-7395-4c32-88c5-9d490be16517">

hxunsc replied:
이때 데이터베이스와 관련하여 오류가 발생하여 500이 발생했던 것 같습니다.. 데이터베이스 문제 해결하니 의도한 바로 나오는 것 같습니다!!

seungy0 commented on src/main/java/gift/controller/FrontCategoryController.java:
```diff
@@ -25,12 +25,13 @@ public FrontCategoryController(CategoryService categoryService) {
             .map(CategoryResponse::from)
-            .collect(Collectors.toList());
-        return ResponseEntity.ok(response);
+            .toList();
```
👍

seungy0 commented on src/main/java/gift/controller/KakaoApiController.java:
```diff
@@ -40,21 +49,38 @@ public ResponseEntity<String> getKakaoLoginUrl() {
+        String kakaoEmail = userInfo.getId() + "@kakao.com"; // 사용자 ID 기반 임시 이메일 생성
 
-        return ResponseEntity.ok(userInfo);
+        // 사용자 정보로 회원 가입 또는 로그인 처리
+        Optional<Member> existingMember = memberService.findByEmail(kakaoEmail);
+
+        if (existingMember.isPresent()) {
+            // 기존 회원 -> 로그인 처리
+            Member member = existingMember.get();
+            member.setRefreshToken(tokenResponse.getRefreshToken());
+            memberService.registerNewMember(member);
+            String jwtToken = jwtUtil.createToken(member.getEmail());
+            return ResponseEntity.ok("Bearer " + jwtToken);
+        } else {
+            // 신규 회원 -> 회원 가입 처리
+            Member newMember = new Member(kakaoEmail, "");
+            newMember.setRefreshToken(tokenResponse.getRefreshToken());
+            memberService.registerNewMember(newMember);
+            String jwtToken = jwtUtil.createToken(newMember.getEmail());
+            return ResponseEntity.ok("Bearer " + jwtToken);
+        }
```
이 부분은 컨트롤러의 책임이 아닌 것 같습니다.

seungy0 commented on src/main/java/gift/service/OrderService.java:
```diff
@@ -20,37 +27,66 @@ public class OrderService {
 
     public OrderService(OrderRepository orderRepository, OptionRepository optionRepository,
-        WishRepository wishRepository, OrderMessageService orderMessageService) {
+        WishRepository wishRepository, OrderMessageService orderMessageService, JwtUtil jwtUtil,
+        MemberRepository memberRepository, KakaoService kakaoService) {
         this.orderRepository = orderRepository;
         this.optionRepository = optionRepository;
         this.wishRepository = wishRepository;
         this.orderMessageService = orderMessageService;
+        this.jwtUtil = jwtUtil;
+        this.memberRepository = memberRepository;
+        this.kakaoService = kakaoService;
```
서비스가 점점 무거워지고 있습니다. 어떻게 해결할 수 있을까요?
일단 jwtUtil은 유틸 클래스이므로 정적 메서드로 의존성 주입 없이 사용하는 것도 좋을 것 같습니다.

hxunsc replied:
현재 JwtUtil은 정적 메서드로 사용하도록 수정하였습니다.. 근데 그 외의 부분은 아직 고민중인 상태입니다.. 서비스단을 더 분리하는 것이 좋을까요?

seungy0 commented on src/main/java/gift/service/OrderService.java:
```diff
@@ -20,37 +27,66 @@ public class OrderService {
-        orderMessageService.sendOrderMessage(order, token);
+
+        // 새로운 Access Token 발급
+        KakaoAccessToken newAccessToken = kakaoService.refreshAccessToken(member.getRefreshToken());
```
리프레시 토큰이 만료되었다면 어떻게 할까요?

hxunsc replied:
만료 시 관련 에러코드를 던지도록 수정했습니다!

seungy0 reviewed (approved):
현서님 고생하셨습니다~
